tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.py

```diff
@@ -0,0 +1,267 @@
+
+      self.assertItemsEqual(full_filenames, produced_filenames)
+
+      with self.assertRaises(errors.OutOfRangeError):
+        sess.run(itr.get_next())
+
+  def testFileSuffixes(self):
+    filenames = ['a.txt', 'b.py', 'c.py', 'd.pyc']
+    self._touchTempFiles(filenames)
+
+    filename_placeholder = array_ops.placeholder(dtypes.string, shape=[])
+    dataset = MatchingFilesDataset(filename_placeholder)
+
+    with self.cached_session() as sess:
+      itr = iterator_ops.Iterator.from_structure(dataset.output_types)
+      init_op = itr.make_initializer(dataset)
+      next_element = itr.get_next()
+      sess.run(
+        init_op,
+        feed_dict={filename_placeholder: path.join(self.tmp_dir, '*.py')})
+
+      full_filenames = []
+      produced_filenames = []
+      for filename in filenames[1:-1]:
+        full_filenames.append(
+          compat.as_bytes(path.join(self.tmp_dir, filename)))
+        produced_filenames.append(compat.as_bytes(sess.run(next_element)))
+      self.assertItemsEqual(full_filenames, produced_filenames)
+
+      with self.assertRaises(errors.OutOfRangeError):
+        sess.run(itr.get_next())
+
+  def testFileMiddles(self):
+    filenames = ['a.txt', 'b.py', 'c.pyc']
+    self._touchTempFiles(filenames)
+
+    filename_placeholder = array_ops.placeholder(dtypes.string, shape=[])
+    dataset = MatchingFilesDataset(filename_placeholder)
+
+    with self.cached_session() as sess:
+      itr = iterator_ops.Iterator.from_structure(dataset.output_types)
+      init_op = itr.make_initializer(dataset)
+      next_element = itr.get_next()
+      sess.run(
+        init_op,
+        feed_dict={filename_placeholder: path.join(self.tmp_dir, '*.py*')})
+
+      full_filenames = []
+      produced_filenames = []
+      for filename in filenames[1:]:
+        full_filenames.append(
+          compat.as_bytes(path.join(self.tmp_dir, filename)))
+        produced_filenames.append(compat.as_bytes(sess.run(next_element)))
+
+      self.assertItemsEqual(full_filenames, produced_filenames)
+
+      with self.assertRaises(errors.OutOfRangeError):
+        sess.run(itr.get_next())
+
+  def _load_data(self):
+    new_files = []
+    dir = "/tmp/test/"
+    if not os.path.exists(dir):
+      os.makedirs(dir)
+    base = tempfile.mkdtemp(prefix=dir)
+    print('saving files to dir: {}'.format(base))
+    for i in range(width):
+      new_base = os.path.join(base, str(i), *[str(j) for j in range(depth - 1)])
+      if not os.path.exists(new_base):
+        os.makedirs(new_base)
+      f = os.path.join(new_base, 'stuff.txt')
+      new_files.append(compat.as_bytes(f))
+      open(f, 'w').close()
+    return base, new_files
+
+  def _read_data(self, data, sess, N=1):
+    for _ in range(N):
+      sess.run(data)
+
+  def _read_data_with_result(self, data, sess, N=1):
+    result = []
+    for _ in range(N):
+      result.append(sess.run(data))
+    return result
+
+  def testPerformance(self):
```

Comment:
This should be a benchmark. See `map_dataset_op_test.py` for an example of how to write one.